File: c/src/Board.c
/*
 * Board.c
 *
 *  Created on: Mar 30, 2017
 *      Author: luke
 */

#include <stdio.h>
#include "Board.h"
#define NO_PIECE -1

#define VALIDATE_POSITION(coordinate,op) \
		if (!isValidPosition(board, coordinate)) { \
			fprintf(stderr, "Cannot %s piece at %d,%d as the location is invalid",#op,coordinate.row, coordinate.col); \
			return NULL; \
		}

#define MAKE_NO_PIECE(row, col) \
		board->pieces[row][col].color = NO_PIECE

int isFinalRowForPiece(Piece *piece, Coordinate dest) {
	return (piece->color == RED && dest.row == 0) || (piece->color == BLACK && dest.row == (BOARD_SIZE-1));
}

void moveAndKingPiece(Piece *piece, Board *board, Coordinate dest) {
	//printf("Board.c: moving piece to %d,%d: ", dest.row, dest.col); printPiece(piece);
	if (isFinalRowForPiece(piece, dest) && piece->canBeKinged) {
		makeKingPiece(getPieceAt(board, dest), piece->color);
	}
	else {
		placePieceAt(board, piece, dest);
	}
}
void removePieceAt(Board *board, Coordinate coordinate) {
	//printf("Board.c: Removing piece 0x%x from %d,%d\n", &(board->pieces[coordinate.row][coordinate.col]), coordinate.row, coordinate.col);
	MAKE_NO_PIECE(coordinate.row,coordinate.col);
}

void placePieceAt(Board *board, Piece *piece, Coordinate coordinate) {
	//printf("Board.c: Placing piece %d on %d,%d\n", piece->color, coordinate.row, coordinate.col);
	board->pieces[coordinate.row][coordinate.col] = *piece;
	//printPiece(piece);
}

PieceList getPiecesForColor(Board *board, color_t color) {
	PieceList pieces;
	pieces.count = 0;
	for (unsigned row = 0; row < BOARD_SIZE; row++) {
		for (unsigned col = 0; col < BOARD_SIZE; col++) {
			Piece piece = board->pieces[row][col];
			if (piece.color == color) {
				if (pieces.count == 12) {
					fprintf(stderr, "WTF WE HAVE 13 pieces on the board!!!!\n");
					printBoard(board);
				}
				pieces.pieces[pieces.count].coordinate = (Coordinate){.row=row, .col=col};
				pieces.pieces[pieces.count++].piece = piece;
			}
		}
	}
	return pieces;
}


void initialize(Board *board) {
	for (unsigned row = 0; row < BOARD_SIZE; row++) {
		for (unsigned col = 0; col < BOARD_SIZE; col++) {
			MAKE_NO_PIECE(row,col);
		}
	}
}

int isPiece(Piece *piece) {
	return piece->color != NO_PIECE;
}

int isEven(unsigned i) {
	return i % 2 == 0;
}

int isUsableSquare(unsigned row, unsigned col) {
	return (isEven(row) && isEven(col)) || (!isEven(row) && !isEven(col));
}

int isValidPosition(Board *board, Coordinate c) {
	return 	c.row >= 0 &&
			c.row < BOARD_SIZE &&
			c.col >= 0 &&
			c.col < BOARD_SIZE &&
			isUsableSquare(c.row, c.col);
}

void printPiece(Piece *piece);
int pieceIsJumpable(Board *board, Piece *piece, Coordinate coordinate) {
	Piece *otherPiece = getPieceAt(board, coordinate);
	//printf("Board.c::PIJ Piece is "); printPiece(piece);
	//printf("\tOther piece is "); printPiece(otherPiece);
	//printf("\tAnd isPiece is %d and colorTest is %d\n", isPiece(otherPiece), otherPiece->color != piece->color);
	//printf("\tat %d,%d\n",coordinate.row, coordinate.col);
	return isPiece(otherPiece) && otherPiece->color != piece->color;
}

int nextSpaceIsLandable(Board *board, Piece *piece, Coordinate from, Coordinate to) {
	Coordinate landingSpot = EXTENDING_COORDINATE(from, to);
	return isValidPosition(board, landingSpot) && !isPiece(getPieceAt(board, landingSpot));
}

Piece *getPieceAt(Board *board, Coordinate coordinate) {
	VALIDATE_POSITION(coordinate, "get");
	return &(board->pieces[coordinate.row][coordinate.col]);
}

void printBoard(Board *board) {
	printf("\n  ");
	for (unsigned col = 0; col < BOARD_SIZE; col++) {
		printf("%d", col);
	}
	printf("\n");
	for (unsigned row = 0; row < BOARD_SIZE; row++) {
		printf("%d ", row);
		for (unsigned col = 0; col < BOARD_SIZE; col++) {
			if (!isUsableSquare(row,col)) {
				printf("|");
			}
			else {
				Piece *p = &(board->pieces[row][col]);
				if (!isPiece(p)) { printf("_"); }
				else if (p->color == BLACK) {
					if (p->canBeKinged) { printf("b"); }
					else { printf("B"); }
				}
				else if (p->color == RED) {
					if (p->canBeKinged) { printf("r"); }
					else { printf("R"); }
				}
				else { printf("?"); }
			}
		}
		printf("\n");
	}
}

File: python/python/src/Board.py
'''
Created on Apr 16, 2017

@author: luke
'''

from color import Color

class Board(object):
    '''
    classdocs
    '''
    DEFAULT_SIZE = 8

    class PlacedPiece(object):
        def __init__(self, piece, coordinate):
            if isinstance(piece, Board.PlacedPiece):
                raise Exception("PASSING IN A PLACED PIECE TO A PLACED PIECE")
            self.piece = piece
            self.coordinate = coordinate
            self.getSimpleMoves = lambda b: piece.getSimpleMoves(self.coordinate, b)
            #self.getJumpMoves = lambda b: piece.getJumpMoves(self.coordinate, b)
        
        def __str__(self):
            return str(self.piece) + "@" + str(self.coordinate)
        
        def __repr__(self):
            return str(self.piece) + "@" + str(self.coordinate)
        
        def __getattr__(self, name):
            theMethod = getattr(self.piece, name)
            if theMethod is not None:
                return theMethod
            else:
                print ("UNSUPPORTED GET: " + name)
                
        def getJumpMoves(self, board):
            try:
                return self.piece.getJumpMoves(self.coordinate, board)
            except Exception as e:
                print("BAD! Piece is " + self.piece)
            
            
    def __init__(self, size=DEFAULT_SIZE):
        self.squares = [[None for x in range(size)] for y in range(size)]
        
    def size(self):
        return len(self.squares)
    
    def isValidPosition(self, coordinate):
        return coordinate.row >= 0 and coordinate.col >= 0\
                and coordinate.row < len(self.squares) and coordinate.col < len(self.squares)\
                and self.isUsableSquare(coordinate.row, coordinate.col)
    
    def isUsableSquare(self, row, col):
        return (isEven(row) and isEven(col)) or (not isEven(row) and  not isEven(col))
    
    def placePieceAt(self, piece, coordinate):
        if isinstance(piece, Board.PlacedPiece):
            placed = piece
        else:
            placed = Board.PlacedPiece(piece,coordinate)
        self.squares[coordinate.row][coordinate.col] = placed
        return placed
    
    def getPiecesForColor(self, color):
        pieces = []
        for row in range(len(self.squares)):
            pieces.extend([piece for piece in self.squares[row] if piece is not None and piece.color == color])
        return pieces
    
    def isFinalRowForPiece(self, piece, coordinate):
        return (piece.color == Color.RED and coordinate.row == 0) \
            or (piece.color == Color.BLACK and coordinate.row == len(self.squares)-1)
    
    def __getitem__(self, coordinate):
        return self.squares[coordinate.row][coordinate.col]
    
    def __delitem__(self, coordinate):
        self.squares[coordinate.row][coordinate.col] = None
    
    def __setitem__(self, coordinate, piece):
        self.placePieceAt(piece, coordinate)
        
    def __repr__(self):
        s = "  "
        s += "".join(map(str,range(len(self.squares))))
        s += "\n"
        for row in range(len(self.squares)):
            s += str(row) + " "
            for col in range(len(self.squares)):
                if not self.isUsableSquare(row,col):
                    s += "|"
                else:
                    p = self.squares[row][col]
                    if p is None:
                        s += "_"
                    elif p.color == Color.BLACK:
                        if p.canBeKinged:
                            s += "b"
                        else:
                            s += "B"
                    elif p.color == Color.RED:
                        if p.canBeKinged:
                            s += "r"
                        else:
                            s += "R"
                    else:
                        s += "?"
            s += "\n"
        return s
    
def isEven(num):
    return num % 2 == 0